js_modules/dagster-ui/packages/ui-core/src/runs/RawLogContent.tsx

```diff
@@ -253,18 +253,35 @@ class ScrollContainer extends React.Component<IScrollContainerProps> {
 
 const LineNumbers = (props: IScrollContainerProps) => {
   const {content} = props;
-  if (!content) {
-    return null;
-  }
-  const matches = content.match(/\n/g);
+  const lastCount = React.useRef(0);
+  const container = React.createRef<HTMLDivElement>();
+
+  const matches = (content || '').match(/\n/g);
   const count = matches ? matches.length : 0;
-  return (
-    <LineNumberContainer>
-      {Array.from(Array(count), (_, i) => (
-        <div key={i}>{String(i + 1)}</div>
-      ))}
-    </LineNumberContainer>
-  );
+
+  // The common case here is 1+ new line numbers appearing on each render. Until we fully
+  // virtualize this UI, a good solution is to append a new div containing just the added
+  // line numbers. This avoids repaint + relayout of the existing line numbers, which takes
+  // 100ms per 100k lines of logs.
+  React.useLayoutEffect(() => {
+    const containerEl = container.current;
+    if (!containerEl) {
+      return;
+    }
+    if (count < lastCount.current) {
+      containerEl.textContent = '';
+      lastCount.current = 0;
+    }
+    const div = document.createElement('div');
+    const addedCount = count - lastCount.current;
+    div.textContent = Array.from(Array(addedCount), (_, i) =>
+      String(lastCount.current + i + 1),
+    ).join('\n');
+    containerEl.appendChild(div);
```

Comment:
Could we end up with the line numbers misaligned with respect to the log lines? (Of course I don't know whether that's already a problem...)

Reply:
Yeah I'm actually really surprised this was working at all -- it does seem to work correctly but only because we don't do text wrapping on the raw logs, and each line has a consistent height